### PR TITLE
DRV-490: Streaming

### PR DESCRIPTION
**Pull request by parkhomenko**
[Notes]
Introduces streaming for .net driver

[Jira ticket]
https://faunadb.atlassian.net/browse/DRV-490

### CODE REVIEW

**Comment by n400** [comment]
Could we release this to Preview first to see if any of our C# users are willing to help us test it?

**Comment by faunaee** [comment]
@n400 Drivers aren't released to Preview/Production, they are released to the host language's package site. For the C# driver, we'd likely have to release a "preview" package to NuGet, and tell users how to connect to the Preview environment.

**Comment by n400** [comment]
> @n400 Drivers aren't released to Preview/Production, they are released to the host language's package site. For the C# driver, we'd likely have to release a "preview" package to NuGet, and tell users how to connect to the Preview environment.

Yes, I understand as that is what happened I got testers for the javascript API v4 preview driver. Releasing "to Preview" is just shorthand. I'm asking if we can release a preview version of the driver like we did for the javascript driver.

File: FaunaDB.Client/Client/StreamingEventHandler.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        private FaunaException constructStreamingException(Exception ex)
+        {
+            var queryError = new QueryError(null, "internal exception", ex.Message, null);
+            var response = new QueryErrorResponse(500, new List<QueryError> {queryError});
+            return new StreamingException(response);
+        }
+
+        private FaunaException constructStreamingException(ObjectV value)
+        {
+            var code = value.At("event", "code").To<string>().Value;
```
Comment (marrony) [inline FaunaDB.Client/Client/StreamingEventHandler.cs:79]: create a class Field for this, `var Code = Field.At("event", "code").To<String>()`

File: FaunaDB.Client/Client/StreamingEventHandler.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                observer.OnCompleted();
+            }
+        }
+
+        public void Dispose()
+        {
+            streamReader.Dispose();
+        }
+
+        private FaunaException constructStreamingException(Exception ex)
```
Comment (marrony) [inline FaunaDB.Client/Client/StreamingEventHandler.cs:70]: nit pick, c# methods starts with upper case

**Review by marrony — COMMENTED** [review]
I think Streaming tests deserves its own test file

File: FaunaDB.Client/Client/StreamingEventHandler.cs (hunk 1 above)
Comment (parkhomenko) [reply FaunaDB.Client/Client/StreamingEventHandler.cs]: created fields for `code`, `description` and `type`

File: FaunaDB.Client/Client/StreamingEventHandler.cs (hunk 2 above)
Comment (parkhomenko) [reply FaunaDB.Client/Client/StreamingEventHandler.cs]: sure, missed that one...

**Comment by parkhomenko** [comment]
> I think Streaming tests deserves its own test file

moved streaming tests into a separate file

**Comment by evbruno** [comment]
Is there any new restrictions for running the streaming tests from macOS?
```
 FaunaDB.Client -> /Users/ed/fauna/faunadb-csharp/FaunaDB.Client/bin/Debug/net45/FaunaDB.Client.dll
/Users/ed/fauna/faunadb-csharp/FaunaDB.Client.Test/StreamingTest.cs(3,26): error CS0234: The type or namespace name 'Immutable' does not exist in the namespace 'System.Collections' (are you missing an assembly reference?) [/Users/ed/fauna/faunadb-csharp/FaunaDB.Client.Test/FaunaDB.Client.Test.csproj]
```

**Comment by parkhomenko** [comment]
> Is there any new restrictions for running the streaming tests from macOS?
> 
> ```
>  FaunaDB.Client -> /Users/ed/fauna/faunadb-csharp/FaunaDB.Client/bin/Debug/net45/FaunaDB.Client.dll
> /Users/ed/fauna/faunadb-csharp/FaunaDB.Client.Test/StreamingTest.cs(3,26): error CS0234: The type or namespace name 'Immutable' does not exist in the namespace 'System.Collections' (are you missing an assembly reference?) [/Users/ed/fauna/faunadb-csharp/FaunaDB.Client.Test/FaunaDB.Client.Test.csproj]
> ```

Don't think there are any new restrictions. I'm not an expert in .NET, so I could have done something wrong.

I used a namespace `System.Collections.Immutable` which is included in `.netcore 3.x` and `.net 5`, but not for `.net framework 4.5`.
I've removed it and changed it to a simple `List`, so you shouldn't have this error anymore.

I run the tests against:
- .net core 3.0 [succeeded]
- .net core 2.0 [succeeded]
- .net core 1.1 [succeeded]
- .net core 1.0 [failed] (not only my tests are failing here, but there are also lots of dependencies that are not supported)
- .net 4.5 (since I'm using mac, I used mono to test this one) [succeeded]

Do we need to support all the frameworks specified in this file to run tests?
https://github.com/fauna/faunadb-csharp/blob/master/FaunaDB.Client.Test/FaunaDB.Client.Test.csproj#L16

**Comment by evbruno** [comment]
`Passed!  - Failed:     0, Passed:   308, Skipped:     0, Total:   308, Duration: 19 s - /Users/ed/fauna/faunadb-csharp/FaunaDB.Client.Test/bin/Debug/net45/FaunaDB.Client.Test.dll (net45)`

lgtm !

**Review by evbruno — APPROVED** [review]

**Review by marrony — APPROVED** [review]

**Pull request merged**